Added partial visualisation for Audio Datasets under tfds

harshitadd opened the pull request:
Added partial visualization support for audio datasets under tfds: Tested on LJspeech, Groove
Issue Link: [1528](https://github.com/tensorflow/datasets/issues/1528)

Procedure:
1. Pick random samples from argument dataset 
2. Trim the picked audio samples to include only the first 6 seconds of the audio 
3. Use IPython to display the trimmed audio sample 
4. Plot the corresponding audio waveforms. 

Notebook demonstration: [Link](https://colab.research.google.com/drive/11NSYCRPiHsDCQZBxDVew-OniGbscdwxH)

Eshan-Agarwal commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -66,17 +70,13 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
-      raise ValueError(
-          "Visualisation not supported for dataset `{}`. Was not able to "
-          "auto-infer image.".format(ds_info.name))
-
```
Why removing these, I think we should print a message if provided data is video or text. Maybe you can add param like audio key and use `if not image_keys and not audio_keys` then again check for `if not image_keys` so print this message else print shows error like visualization is supports only for audio and mage dataset

Eshan-Agarwal commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -66,17 +70,13 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
   plt = lazy_imports_lib.lazy_imports.matplotlib.pyplot
 
   if not image_key:
```
Also use `if not image_key and not audio_key` after adding audio key param and generate audio_keys using `features_lib.Audio` so that if any other dataset like video or text is passed it can print error

harshitadd replied:
I tried this - feature_lib.Audio was returning None for both the tested datasets. Since that instantiation check was not working - I had to resort to this implementation. [][Here](https://github.com/tensorflow/datasets/pull/1683/commits/cfb259b8ad3b828935529f341cd9b1e2a86fd103)) is the commit with that code - Perhaps you could point out the mistake?

Eshan-Agarwal replied:
Don't know why its not working , dataset you are trying for audio_keys have how many number of channels, Audio feature supports only 1-D values [see](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/core/features/audio_feature.py#L47)

harshitadd replied:
Just rechecked - I think I was missing something trying it earlier - the instantiation check is working (at least with the feature_lib.Audio) so I will add that. My bad.  
Thanks for the code style check reminder - have done that as well and will make the required changes! :+1:

Eshan-Agarwal commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -85,46 +85,82 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
 
     image_key = image_keys[0]
```
I think check first like` if image_keys` then only set `image_key `as if data is not image dataset `image_key` set to empty

harshitadd replied:
Refer to comment 78 and code 79 - The check to run this code has been added ( only when the image instances are populated )

Eshan-Agarwal replied:
Ohh sorry its my mistake

Eshan-Agarwal commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -85,46 +85,82 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
+    if not label_key:
+      logging.info("Was not able to auto-infer label.")
+
+    num_examples = rows * cols
```
Add a check here `if image_key` so that below code runs only if data is image dataset

harshitadd replied:
Refer to code line 78 - This is all under if image keys: indicating that the code runs only when the image keys have been successfully generated.

Eshan-Agarwal commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -85,46 +85,82 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
+    return fig
+
+  # if not image item instances 
+  if not image_key:
```
What if dataset is like video dataset then it tries to run this command also so change it to `if audio_key` after passing audio_key param

harshitadd replied:
Agreed - As mentioned - The audio key instantiation check wasn't working ( commit linked above ) - I am working on resolving that, which should allow introducing video/text/audio based instance checks.

Eshan-Agarwal replied:
Yes please

Conchylicultor commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -20,6 +20,10 @@
 from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
```
Internally, we are still supporting Python2, so it shouldn't be removed yet.

Eshan-Agarwal replied:
@Conchylicultor  thanks for correction

Eshan-Agarwal replied:
@Conchylicultor  thanks for correction

Eshan-Agarwal commented:
@harshitadd  Thanks for quick changes, Also please make sure if we provide 2d audio data it displays error ? and finally again run `./oss_scripts/lint.sh tensorflow_datasets/image`

Conchylicultor commented:
Thank you for your contribution.

harshitadd commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -30,21 +30,16 @@
 
 def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
   """Visualize images (and labels) from an image classification dataset.
-
```
Will bear in mind - I had omitted them while checking formatting errors generated by running oss_scripts/lint.sh.

harshitadd commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -128,3 +128,36 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
       plt.xlabel("{} ({})".format(label_str, label))
   plt.show()
   return fig
+  if audio_keys:
```
Sure!

harshitadd commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -128,3 +128,36 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
       plt.xlabel("{} ({})".format(label_str, label))
   plt.show()
   return fig
+  if audio_keys:
```
Noted, Should I run the code with other types of datasets or add a test case explicitly in show_examples_test.py ?

harshitadd commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -128,3 +128,36 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
   return fig
+  if audio_keys:
+    audio_samples = []
+    if(ds_info.name == 'ljspeech'):
```
Will do.

harshitadd commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -128,3 +128,36 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
+    for audio in to_gen:
+      name = '/content/audio.wav'
+      newAudio = audio[t1:t2]
+      write(name, samplerate, newAudio)
```
Imported in the beginning as scipy.io.wavfile.write

harshitadd commented on tensorflow_datasets/core/visualization.py:
```diff
@@ -128,3 +128,36 @@ def show_examples(ds_info, ds, rows=3, cols=3, plot_scale=3., image_key=None):
+      name = '/content/audio.wav'
+      newAudio = audio[t1:t2]
+      write(name, samplerate, newAudio)
+      IPython.display.display(IPython.display.Audio(name)) 
```
Working on it, Thank you so much for your comments.

harshitadd commented:
@Conchylicultor - I tried refactoring the code to plug it in the visualizer codebase but the audio_key extraction seems to be returning a Null. 
I have removed the write on disk dependency ( audio now displays directly from the np array ) and added the missing imports.

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,68 @@
+
+
+import random
+import IPython.display
```
This will crash for all user not running in a non-ipython environement. This should not be imported in the global scope

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,68 @@
+""" Audio Visualization
```
Could you add a test in `show_example_test.py` to make sure this works ? You can use https://docs.python.org/3/library/unittest.mock.html to make sure the AudioVisualizer is chosen.

harshitadd replied:
I tried running show_example_test ( for the imagenet data only ) and encounter the following error : 

ValueError: TFDS has been mocked, but metadata files were not found in /usr/local/lib/python3.6/dist-packages/tensorflow_datasets/testing/metadata/imagenet2012/5.0.0. You should copy the real metadata files, so that the dataset can be loaded properly, or set the data_dir kwarg oftfds.testing.mock_tfds(data_dir=...).

----------------------------------------------------------------------
Ran 2 tests in 0.540s

FAILED (errors=1, skipped=1)

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,68 @@
+  ):
+    """Display the dataset.
+
+      Args:
```
Arguments are not matching

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,68 @@
+        image_key: `string`, name of the feature that contains the image. If not
+          set, the system will try to auto-detect it.
+      """
+    key = audio_keys[0]
```
Where is it declared ?

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,75 @@
+        image_key: `string`, name of the feature that contains the image. If not
+          set, the system will try to auto-detect it.
+      """
+    import random
```
Random is a standard Python module, so installed in every users, so it is safe to keep it in the global scope

harshitadd commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,68 @@
+from tensorflow_datasets.core import features as features_lib
+from tensorflow_datasets.core import lazy_imports_lib
+from tensorflow_datasets.core.visualization import visualizer
+plt = lazy_imports_lib.lazy_imports.matplotlib.pyplot
```
> To check if a dataset has an audio feature, you can see on our catalog: For instance https://www.tensorflow.org/datasets/catalog/crema_d
> 
> It seems Groove is using `tfds.features.Tensor` instead of `tfds.features.Audio`, which sounds like a bug to me, we should upgrade groove to use audio feature. I'll open a bug for this:
> https://www.tensorflow.org/datasets/catalog/groove#groovefull-16000hz

Thank you for your comments - A bug for the Groove 'Tensor' attribute will be helpful - Should I report it as a bug ?

Conchylicultor replied:
The bug has been fixed so it should works if you are rebasing from master.

harshitadd replied:
It does, Thanks! The auto inferring works now ( [Link](https://colab.research.google.com/drive/1cK8De-bi3E87-LifRodsuAkvMpu1BLV-) . I have tested it with Groove, crema_d, ljspeech.

harshitadd commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,68 @@
+from tensorflow_datasets.core.visualization import visualizer
+plt = lazy_imports_lib.lazy_imports.matplotlib.pyplot
+
+class AudioGridVisualizer(visualizer.Visualizer):
```
Will do!

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,69 @@
+    key = audio_keys[0]
+    audio_samples = []
+
+    samplerate = 16000
```
Now you can use `ds_info.features[key].sample_rate` when defined (and use default value if `sample_rate is None`)

harshitadd replied:
Added

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,69 @@
+      audio = audio[t1:t2]
+      ipd.display(ipd.Audio(audio, rate=samplerate))
+    plt = lazy_imports_lib.lazy_imports.matplotlib.pyplot
+    fig, a = plt.subplots(2, 2)
```
To make the code more generic, you could refactor the code to use `rows` and `cols`, similarly to  https://github.com/tensorflow/datasets/blob/b5d7ec65b84aa95e7f5e78f01f7698958498f65c/tensorflow_datasets/core/visualization/image_visualizer.py

And ideally, you could try to reuse the `make_grid` function https://github.com/tensorflow/datasets/blob/b5d7ec65b84aa95e7f5e78f01f7698958498f65c/tensorflow_datasets/core/visualization/image_visualizer.py#L31
Or make a similar util function.

harshitadd replied:
All Right! I made some edits accordingly

Conchylicultor commented on tensorflow_datasets/core/visualization/audio_visualizer.py:
```diff
@@ -0,0 +1,69 @@
+      ds,
+      rows=2,
+      cols=2,
+      plot_scale=3.,
```
Some of the arguments here are unused.
